YubiKit/YubiKit/Helpers/YKFTLVRecord.m

```diff
@@ -0,0 +1,166 @@
+// Copyright 2018-2022 Yubico AB
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#import <Foundation/Foundation.h>
+#import "YKFTLVRecord.h"
+#import "YKFNSDataAdditions+Private.h"
+
+@interface YKFTLVRecord()
+@property (nonatomic, readwrite) YKFTLVTag tag;
+@property (nonatomic, readwrite) NSData *value;
+@end
+
+@interface NSData (NSData_ByteEncoding)
+    + (NSData *)ykf_dataWithBytesStripLeadingZeros:(char*)bytes length:(int)length;
+@end
+
+@implementation YKFTLVRecord
+
++ (nullable instancetype)recordFromData:(NSData *_Nullable)data checkMatchingLength:(Boolean)checkMatchingLength bytesRead:(int*)bytesRead {
+    *bytesRead = 0;
+    
+    // tag
+    if (data.length == 0) {
+        return nil;
+    }
+    Byte *bytes = (Byte *)data.bytes;
+    int offset = 0;
+    int tag = bytes[offset++] & 0xFF;
+    if ((tag & 0x1F) == 0x1F) {
+        tag = (tag << 8) | (bytes[offset++] & 0xFF);
+        while ((tag & 0x80) == 0x80 && offset < data.length) {
+            tag = (tag << 8) | (bytes[offset++] & 0xFF);
```

Comment:
Should be & 0x1f as the high bit just indicates that it's not the last byte

Reply:
There should also be a failure if the tag exceeds the data, now it will just stop reading it

Reply:
And the test for the high bit must use bytes[offset -1] since it's not present in tag now

Reply:
So I'd recommend saving it like sp:
Byte last_read = bytes[offset++]